src/Hadrian/Haskell/Cabal/Parse.hs

```diff
@@ -8,36 +8,287 @@
+  , parseCabalPkgId
+  , configurePackage, copyPackage, registerPackage
+  ) where
 
 import Data.List.Extra
 import Development.Shake
+import qualified Distribution.ModuleName               as ModuleName
 import qualified Distribution.Package                   as C
 import qualified Distribution.PackageDescription        as C
+import qualified Distribution.PackageDescription.Configuration as C
 import qualified Distribution.PackageDescription.Parsec as C
+import qualified Distribution.Simple.Compiler          as C (packageKeySupported, languageToFlags, extensionsToFlags, compilerInfo)
+import qualified Distribution.Simple.GHC               as GHC
+import qualified Distribution.Simple.Program.Db        as Db
+import qualified Distribution.Simple                   as Hooks (simpleUserHooks, autoconfUserHooks, defaultMainWithHooksNoReadArgs, compilerFlavor, CompilerFlavor(GHC))
+import qualified Distribution.Simple.UserHooks         as Hooks
+import qualified Distribution.Simple.Program.Builtin   as C
+import qualified Distribution.Simple.Utils             as C (findHookedPackageDesc)
+import qualified Distribution.Simple.Program.Types     as C (programDefaultArgs, programOverrideArgs)
+import qualified Distribution.Simple.Configure         as C (getPersistBuildConfig)
+import qualified Distribution.Simple.Build             as C (initialBuildSteps)
+import qualified Distribution.Types.ComponentRequestedSpec as C (defaultComponentRequestedSpec)
+import qualified Distribution.InstalledPackageInfo as Installed
+import qualified Distribution.Simple.PackageIndex as PackageIndex
+import qualified Distribution.Types.LocalBuildInfo as C
 import qualified Distribution.Text                      as C
 import qualified Distribution.Types.CondTree            as C
+import qualified Distribution.Types.MungedPackageId    as C (mungedName)
 import qualified Distribution.Verbosity                 as C
 
-import Hadrian.Haskell.Cabal.Type
-
--- | Parse a Cabal file.
-parseCabal :: FilePath -> IO Cabal
-parseCabal file = do
-    gpd <- liftIO $ C.readGenericPackageDescription C.silent file
-    let pd      = C.packageDescription gpd
-        pkgId   = C.package pd
-        name    = C.unPackageName (C.pkgName pkgId)
-        version = C.display (C.pkgVersion pkgId)
-        libDeps = collectDeps (C.condLibrary gpd)
-        exeDeps = map (collectDeps . Just . snd) (C.condExecutables gpd)
-        allDeps = concat (libDeps : exeDeps)
-        sorted  = sort [ C.unPackageName p | C.Dependency p _ <- allDeps ]
-        deps    = nubOrd sorted \\ [name]
-    return $ Cabal deps name (C.synopsis pd) version
-
-collectDeps :: Maybe (C.CondTree v [C.Dependency] a) -> [C.Dependency]
-collectDeps Nothing = []
-collectDeps (Just (C.CondNode _ deps ifs)) = deps ++ concatMap f ifs
-  where
-    f (C.CondBranch _ t mt) = collectDeps (Just t) ++ collectDeps mt
+import Base
+import Builder hiding (Builder)
+import Context -- .Type
```

Comment:
To type or not to type? Let's make the decision.

Reply:
Fixed in an upcoming commit.